File: src/model/coerce.ts
import { BeckError } from '../util/errors'
import { accentToCss } from '../util/color'

// Low-level coercion helpers shared by every diagram-type builder. Each throws
// a BeckError with a friendly field path instead of letting a TypeError leak.

export function asObject(v: unknown, field: string): Record<string, unknown> {
  if (v == null) return {}
  if (typeof v !== 'object' || Array.isArray(v)) throw new BeckError(`\`${field}\` must be a mapping`)
  return v as Record<string, unknown>
}

export function asArray(v: unknown, field: string): unknown[] {
  if (v == null) return []
  if (!Array.isArray(v)) throw new BeckError(`\`${field}\` must be a list`)
  return v
}

export function asString(v: unknown, field: string): string {
  if (typeof v === 'string') return v
  if (typeof v === 'number' || typeof v === 'boolean') return String(v)
  throw new BeckError(`\`${field}\` must be a string`)
}

export function optString(v: unknown): string | undefined {
  if (v == null) return undefined
  if (typeof v === 'string') return v
  if (typeof v === 'number' || typeof v === 'boolean') return String(v)
  return undefined
}

export function optColor(v: unknown): string | undefined {
  const s = optString(v)
  return s == null ? undefined : accentToCss(s, 'primary')
}

export function optNumber(v: unknown, field: string): number | undefined {
  if (v == null) return undefined
  if (typeof v === 'number') return v
  if (typeof v === 'string' && v.trim() !== '' && !Number.isNaN(Number(v))) return Number(v)
  throw new BeckError(`\`${field}\` must be a number`)
}

export function optBool(v: unknown, field: string, dflt: boolean): boolean {
  if (v == null) return dflt
  if (typeof v === 'boolean') return v
  if (v === 'true') return true
  if (v === 'false') return false
  throw new BeckError(`\`${field}\` must be true or false`)
}

/** Tri-state boolean: undefined when unset (so a heuristic can decide later). */
export function triBool(v: unknown, field: string): boolean | undefined {
  if (v == null) return undefined
  return optBool(v, field, false)
}

export function oneOf<T extends string>(v: unknown, allowed: readonly T[], field: string, dflt: T): T {
  if (v == null) return dflt
  const s = String(v)
  if ((allowed as readonly string[]).includes(s)) return s as T
  throw new BeckError(`\`${field}\` must be one of: ${allowed.join(', ')} (got "${s}")`)
}

/** Coerce a list of scalars (class fields/methods and similar line lists). */
export function stringList(v: unknown, field: string): string[] {
  return asArray(v, field).map((s) => asString(s, field))
}

File: src/Beck/Model/Coerce.cs
namespace Beck.Model;

/// <summary>
/// Low-level coercion helpers shared by every diagram-type builder — a port of
/// <c>src/model/coerce.ts</c>. Each throws a <see cref="BeckYamlException"/> with a
/// friendly field path instead of letting a raw type error leak. Operates on the
/// untyped <c>object?</c> tree from <see cref="YamlLoader"/>.
/// </summary>
internal static class Coerce
{
    private static readonly Dictionary<string, object?> _emptyMap = new();
    private static readonly List<object?> _emptyList = [];

    public static IReadOnlyDictionary<string, object?> AsObject(object? v, string field)
    {
        if (v is null)
        {
            return _emptyMap;
        }

        if (v is IReadOnlyDictionary<string, object?> d)
        {
            return d;
        }

        throw new BeckYamlException($"`{field}` must be a mapping");
    }

    public static IReadOnlyList<object?> AsArray(object? v, string field)
    {
        if (v is null)
        {
            return _emptyList;
        }

        if (v is IReadOnlyList<object?> l)
        {
            return l;
        }

        throw new BeckYamlException($"`{field}` must be a list");
    }

    public static string AsString(object? v, string field)
    {
        if (v is string s)
        {
            return s;
        }

        if (v is double or bool)
        {
            return Js.Str(v);
        }

        throw new BeckYamlException($"`{field}` must be a string");
    }

    public static string? OptString(object? v)
    {
        if (v is null)
        {
            return null;
        }

        if (v is string s)
        {
            return s;
        }

        if (v is double or bool)
        {
            return Js.Str(v);
        }

        return null; // non-scalars silently dropped
    }

    public static string? OptColor(object? v)
    {
        var s = OptString(v);
        return s is null ? null : Colors.AccentToCss(s, AccentToken.Primary);
    }

    public static double? OptNumber(object? v, string field)
    {
        if (v is null)
        {
            return null;
        }

        if (v is double d)
        {
            return d;
        }

        if (v is string s && Js.TryNumber(s, out var n))
        {
            return n;
        }

        throw new BeckYamlException($"`{field}` must be a number");
    }

    public static bool OptBool(object? v, string field, bool dflt)
    {
        if (v is null)
        {
            return dflt;
        }

        if (v is bool b)
        {
            return b;
        }

        if (v is "true")
        {
            return true;
        }

        if (v is "false")
        {
            return false;
        }

        throw new BeckYamlException($"`{field}` must be true or false");
    }

    /// <summary>Tri-state boolean: null when unset (so a heuristic can decide later).</summary>
    public static bool? TriBool(object? v, string field)
    {
        if (v is null)
        {
            return null;
        }

        return OptBool(v, field, false);
    }

    public static TEnum OneOf<TEnum>(object? v, TokenMap<TEnum> map, string field, TEnum dflt)
        where TEnum : struct, Enum
    {
        if (v is null)
        {
            return dflt;
        }

        var s = Js.Str(v);
        if (map.TryParse(s, out var e))
        {
            return e;
        }

        throw new BeckYamlException(
            $"`{field}` must be one of: {string.Join(", ", map.Tokens)} (got \"{s}\")");
    }

    /// <summary>String variant for transient vocabularies not stored as an enum (e.g. relation kind).</summary>
    public static string OneOfString(object? v, IReadOnlyList<string> allowed, string field, string dflt)
    {
        if (v is null)
        {
            return dflt;
        }

        var s = Js.Str(v);
        if (allowed.Contains(s))
        {
            return s;
        }

        throw new BeckYamlException(
            $"`{field}` must be one of: {string.Join(", ", allowed)} (got \"{s}\")");
    }

    /// <summary>Coerce a list of scalars (class fields/methods and similar line lists).</summary>
    public static List<string> StringList(object? v, string field) =>
        AsArray(v, field).Select(s => AsString(s, field)).ToList();
}